### PR TITLE
Application Import Export for APIM

### PR DESCRIPTION
**Pull request by randilu**
### Proposed changes in this pull request

- This is the Application Import Export feature for WSO2 API Manager.
- This enables users to migrate applications between different environments. 
- Users are able to export an application as a zip archive file, which in return can be imported to a different environment with a different user base.
 

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/test/java/org/wso2/carbon/apimgt/rest/api/store/impl/ExportApiServiceImplTestCase.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
+package org.wso2.carbon.apimgt.rest.api.store.impl;
```
Comment (Krishanx92) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/test/java/org/wso2/carbon/apimgt/rest/api/store/impl/ExportApiServiceImplTestCase.java]: Lets Add License headers for test classes as well

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/test/java/org/wso2/carbon/apimgt/rest/api/store/impl/ImportApiServiceImplTestCase.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
+package org.wso2.carbon.apimgt.rest.api.store.impl;
```
Comment (Krishanx92) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/test/java/org/wso2/carbon/apimgt/rest/api/store/impl/ImportApiServiceImplTestCase.java]: check other places also

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/test/java/org/wso2/carbon/apimgt/rest/api/store/utils/ApplicationImportExportManagerTestCase.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.UUID;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({DAOFactory.class, Application.class})
+public class ApplicationImportExportManagerTestCase {
+
+    private static final Logger log = LoggerFactory.getLogger(ApplicationImportExportManagerTestCase.class);
+
+    APIStore apiStore = Mockito.mock(APIStoreImpl.class);
+    String USER = "admin";
```
Comment (Krishanx92) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/test/java/org/wso2/carbon/apimgt/rest/api/store/utils/ApplicationImportExportManagerTestCase.java]: lets change this variable to private

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+            log.error("Error while exporting Application: " + exportApplication.getName(), e);
+        }
+
+        // Check if no application is exported
+        try {
+            if (APIFileUtils.getDirectoryList(applicationArtifactBaseDirectoryPath).isEmpty()) {
+                // cleanup the archive root directory
+                APIFileUtils.deleteDirectory(path);
+                String errorMsg = "No Applications exported";
+                throw new APIMgtEntityImportExportException(errorMsg, ExceptionCodes.APPLICATION_EXPORT_ERROR);
```
Comment (Krishanx92) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java]: is there any reason for throwing an exception here

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ExportApi.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
+package org.wso2.carbon.apimgt.rest.api.store;
+
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ExportApi.java]: Shall we remove the additional line here?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ImportApi.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
+package org.wso2.carbon.apimgt.rest.api.store;
+
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ImportApi.java]: Same here.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ImportApiServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            applicationDetails.setUpdatedUser(username);
+            ApplicationCreationResponse response = consumer.addApplication(applicationDetails);
+            return Response.status(Response.Status.OK).entity(response).build();
+        } catch (APIManagementException e) {
+            String errorMsg = "Error while importing the Applications";
+            log.error(errorMsg, e);
+            ErrorDTO errorDTO = RestApiUtil.getErrorDTO(e.getErrorHandler());
+            return Response.status(e.getErrorHandler().getHttpStatusCode()).entity(errorDTO).build();
+        }
+
```
Comment (dinusha92) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ImportApiServiceImpl.java]: We can remove extra line here.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ImportApiServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+            return Response.status(Response.Status.OK).entity(updatedApplication).build();
+        } catch (APIManagementException e) {
+            String errorMsg = "Error while importing the Applications";
+            log.error(errorMsg, e);
+            ErrorDTO errorDTO = RestApiUtil.getErrorDTO(e.getErrorHandler());
+            return Response.status(e.getErrorHandler().getHttpStatusCode()).entity(errorDTO).build();
+        }
+
+    }
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ImportApiServiceImpl.java]: Shall we remove unnecessary additional lines in other places as well?

Comment (dinusha92) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ImportApiServiceImpl.java]: Shal we remove extra lines here. Seems they are not needed.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+            throw new APIMgtEntityImportExportException(errorMsg, e);
+        }
+
+        //convert to bean
+        Gson gson = new GsonBuilder().create();
+        //returns an application object from a json string
+        Application applicationDetails = gson.fromJson(applicationDetailsString, Application.class);
+
+        return applicationDetails;
+
```
Comment (dinusha92) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java]: We can remove extra line here.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/test/java/org/wso2/carbon/apimgt/rest/api/store/impl/ExportApiServiceImplTestCase.java (hunk 1 above)
Comment (randilu) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/test/java/org/wso2/carbon/apimgt/rest/api/store/impl/ExportApiServiceImplTestCase.java]: Done

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/factories/ImportApiServiceFactory.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package org.wso2.carbon.apimgt.rest.api.store.factories;
```
[the remark this answers is not in the document]
Comment (randilu) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/factories/ImportApiServiceFactory.java]: Done

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+        } catch (APIManagementException e) {
+            String errorMsg = "Error in accessing uploaded Application archive" + appArchiveLocation;
+            log.error(errorMsg, e);
+            throw new APIMgtEntityImportExportException(errorMsg, e, ExceptionCodes.APPLICATION_IMPORT_ERROR);
+        }
+        Application applicationDetails = parseApplicationFile(archiveExtractLocation);
+        return applicationDetails;
+    }
+
+    private Application parseApplicationFile(String applicationDetailsFilePath)
```
Comment (dinusha92) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java]: Shall we add method comment here?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+                throw new APIMgtEntityImportExportException(errorMsg, ExceptionCodes.APPLICATION_EXPORT_ERROR);
+            }
+        } catch (APIManagementException e) {
+            String errorMsg = "Unable to find Application Details at: " + applicationArtifactBaseDirectoryPath;
+            throw new APIMgtEntityImportExportException(errorMsg, ExceptionCodes.APPLICATION_IMPORT_ERROR);
+        }
+        return applicationArtifactBaseDirectoryPath;
+    }
+    //create a zip archive of the export directory
+    public String createArchiveFromExportedAppArtifacts(String sourceDirectory, String archiveLocation,
```
Comment (dinusha92) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java]: Shall we add a proper method comment here?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java (hunk 11 above)
Comment (randilu) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java]: Done

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java (hunk 12 above)
Comment (randilu) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/utils/FileBasedApplicationImportExportManager.java]: Done

**Pull request merged**